### ISSUE TITLE
Ranger freezes when hovering certain files frequently (macOS)

### ISSUE THREAD

**Issue by tbrodbeck** [ranger/ranger#1787]
#### Runtime Environment

- Operating system and version: macOS 10.15.2 19C57 x86_64
- Terminal emulator and version: iTerm Build 3.3.6
- Python version: 2.7.16
- Ranger version/commit: 1.9.2
- Locale: None.None

#### Current Behavior
Ranger tends to freeze entirely very often. The only help is to press `^c` to interrupt the process and to be able to continue using ranger.

I have not a real understanding what the issue is. 

#### Expected Behavior
Ranger working fluently.

#### Context

I could not get any output with `ranger --debug`. So I still have a hard time to find out, which function is erroneous.

#### Steps to reproduce (for me)
1. `ranger /usr/bin/`
2. Move/scroll down till `applesingle` (This is a default file on macOS, where it always freezes)
3. Try to move further, which is not possible and will only be when hitting `^c`

**Comment by toonn** [comment]
Sounds like some trouble with previews. What does `file applesingle` report the file as?

**Comment by tbrodbeck** [comment]
```
file applesingle                                                                     19:21
applesingle: Mach-O 64-bit executable x86_64
```

Yes its true. When I disable my previews `zp` then it does not freeze.

**Comment by yanzhang0219** [comment]
Same here. Sometimes when I use j/k to go over, suddenly it freezes there. Then I have to press Ctrl-c to restore. As what tbrodbeck did, but not just for applesingle, when I was in /usr/bin/ and I tried to move down, I stuck on /usr/bin/agentxtrap first, and then applesingle, awk, etc..

**Comment by michalens** [comment]
Same issue. Both in kitty and Alacritty.

Perhaps it freezes on large files. If there are files a couple MB large, scroll slows down and when I stop on some file it tries to load it and freezes.

And even if i run `ranger --clean` it still freezes on the same pdf (95KB large) trying to show it as text.

**Comment by toonn** [comment]
@michalens, you could try removing the `pdftotext` line from `scope.sh` and install `mutool` maybe the former's hitting a pathological case here. Is it an OCRed document perhaps?

**Comment by michalens** [comment]
@toonn, It's very difficult to trace the issue. Seems like it's connected with my key repeat speed.
I have a bunch of different files in my downloads folder and when i press and hold `j` it lags and may freeze on any file. Even on some kind of audio file (where preview is just metadata). 
I decreased key repeat speed by one point and now it seems to not freeze, but rather the loading spinner in the top-right corner spins forever. Again, even on files where preview is just metadata.

**Comment by toonn** [comment]
Ah, that's kind of a known (but not understood) issue.

**Comment by unphased** [comment]
I'm seeing this too. Can't really remember for sure (installed Catalina a few days ago) if this issue was also plaguing me on Mojave or not. 

Mainly what I figured out was if I terminate the `ranger/data/scope.sh` child bash process when it is frozen, ranger springs back to life. 

One thing I noticed is that earlier, it was responding to Ctrl+C, but now the lockup is such that SIGINT doesn't do anything to help!

**Comment by unphased** [comment]
> connected with my key repeat speed

Occam's razor would indicate what we're experiencing here to be a garden variety race condition. I'll try to poke around at the scope script to see where it hangs...

**Comment by unphased** [comment]
My instrumentation shows that in all cases the hanging happens on the `bat` call. The shell script just hangs on it. Once I remove it from scope.sh, it no longer freezes. 

I got some problems with this. 1 is the main reason I use ranger is so that I can effortlessly pull up bat previews of files 😛 

2 is bat is not misbehaving by itself. It never hangs when run independently. Also, the frozen state sees the calling bash (scope.sh) hanging with 0% CPU and `bat` has always completed executing by this time.

So it seems like any previewer that takes longer than your key repeat period will trigger some kind of underlying race condition in this preview-caching functionality.

**Comment by unphased** [comment]
Tried a bunch of things... the behavior is that whenever scope.sh calls anything "substantial" (such as bat/pygments) the scope.sh is liable to fail to return for some reason, at least that is the way it looks. This happens even when the scope script is pared down to one single call to pygmentize. Now when bat is used like this instead, it causes a hard freeze, while pygmentize will not freeze it and allows SIGINT to clean up the hung shell. 

The problem lies somewhere upstream of this.

**Comment by unphased** [comment]
Could the problem be related to this comment on the `CommandLoader`?

> Ensure that the process doesn't ever ask for input, otherwise the loader will be blocked until this object is removed from the queue (type ^C in ranger)

I messed around with the stdin argument to Popen but nothing I tried worked. I will just use vim instead of ranger for my use case for the time being.

**Comment by michalens** [comment]
I wonder if adding preview delay (as `set w3m_delay`) for kitty and iTerm2 will fix the issue

**Comment by tbrodbeck** [comment]
> I wonder if adding preview delay (as `set w3m_delay`) for kitty and iTerm2 will fix the issue

Interesting!
But I cannot reproduce my issue right now though, after I reformatted my machine ranger seems to work just fine.

**Comment by michalens** [comment]
> But I cannot reproduce my issue right now though, after I reformatted my machine ranger seems to work just fine.

Does it happen with image previews on? Do you have the same setup? I use kitty and it lags, but when I use iTerm, it's even worse.

**Comment by yanzhang0219** [comment]
The issue still exists. Latest ranger and latest macOS. Tested on the latest alacrity, kitty and iterm2.

**Comment by michalens** [comment]
> How I reproduce this issue quickly: cd into /usr/bin, and the open ranger, and then move the cursor downwards quickly by holding `j`. The cursor will stop at a certain file and then stuck there (freezing). I have to press `ctrl-c` to restore the freezing.

Same. However, if you `set use_preview_script false`, it is very smooth and not freezing.

(Which is just a workaround, because I personally need previews)

P.S. can it be the problem in preview script not being optimized for MacOS?

**Comment by michalens** [comment]
@toonn @yanzhang0219 

If I delete this line from scope.sh - no freezing.
```
local highlight_format='xterm256' 
```

If i understand correctly, it sets Xorg color, which is absent on MacOS.

**Comment by michalens** [comment]
It actually also lags if I set highlight format to 'terminal', 'terminal256' or 'ansi'

PS. I've got no idea what all of this mean, just experimenting. Appreciate, if you could explain.

**Comment by yanzhang0219** [comment]
@michalens Wow, I will try it right now.

**Comment by toonn** [comment]
`xterm256` just means `highlight` will use escape codes for a 256 color terminal. Ideally we'd use `truecolor` but python curses doesn't support that. `ansi` means it sticks to the 16 base terminal colors.  `highlight` defaults to `truecolor` or `xterm256` depending on what's available. So it's weird that removing the option would stop the freezes from happening.

**Comment by yanzhang0219** [comment]
@michalens I tried to comment that line but it also disables preview. For the preview to work normally, I have to set a value to highlight_format, but I have no idea what value I should use.

**Comment by michalens** [comment]
@yanzhang0219 Yes, you are right. 
I put back that line `local highlight_format='xterm256' `,

but now I removed this line:

```
pygmentize -f "${pygmentize_format}" -O "style=${PYGMENTIZE_STYLE}"\                          
-- "${FILE_PATH}" && exit 5     
```

So now text files have previews and seems line no freezing in /usr/bin
@toonn Hope this helps.

However, in folders with image files it still freezes.
Even If I set preview images to false, scrolling is not smooth, because ranges seems to run `exiftool` command on each image which slows it down.
I think adding option to set preview delay would be extremely useful.

**Comment by toonn** [comment]
There is already the `w3m_delay`.

I think this has become more noticeable because of Catalina. Did you update?
Apparently it calls home to apple for any unsigned binaries and scripts that are executed.

**Comment by michalens** [comment]
My preview method is not w3m, but kitty, so it doesn't seem to work.

I am still using Mojave. I briefly tried vifm with image previews and it seems to be very slow too. So, apparently, you are right - it has to do with MacOS. But hopefully there is a workaround

**Comment by yanzhang0219** [comment]
@michalens Thank you!

> but now I removed this line:
> pygmentize -f "${pygmentize_format}" -O "style=${PYGMENTIZE_STYLE}"\                          
> -- "${FILE_PATH}" && exit 5  

Does this removal have any side effects for ranger and its other functionalities? Thanks~

**Comment by andrewferrier** [comment]
> Ah, that's kind of a known (but not understood) issue.

Do you have a link to this issue? I see this too on MacOS - not the hanging, but the perpetual 'spinning' of the ranger spinner when I move too fast over certain documents - for example, `.png` files (which my `scope.sh` uses `exiftool` to introspect, and which display find if I navigate over the file list slowly, or run `scope.sh` independently of ranger).

**Comment by marcelolaia** [comment]
Here I found an issue like that one.
I use ranger with mutt. When I will do a file attach, when I move over a pdf file, a text file is generated side by side the original pdf. When I move down, over the new text file generated form the previous pdf, ranger crash! I have a lot of text file with the same name off other files (i.e. odt, pdf) in my directories.

**Comment by toonn** [comment]
@marcelolaia, that's strange behavior. Could you share the `pdf)` case from your `scope.sh`? The default `scope.sh` we provide does not generate any text files for pdf previews.

**Comment by hntee** [comment]
I just had the same issue here. When pressing `j` and after a few seconds, ranger will hang.

After commenting out these lines in `scope.sh`
```
            #env COLORTERM=8bit bat --color=always --style="plain" \
                #-- "${FILE_PATH}" && exit 5
            #pygmentize -f "${pygmentize_format}" -O "style=${PYGMENTIZE_STYLE}"\
                #-- "${FILE_PATH}" && exit 5
```
the problem seemed to disappear.
But the side effect is the syntax highlighting is also disabled in preview.

Any other workarounds?

**Comment by toonn** [comment]
@hntee, since you didn't remove the `highlight` command for syntax highlighting you could simply install `highlight`.

I suspect the issue is broader though, Catalina calls home to check for infected files, this affects performance to the point of unusability.

**Comment by SergeyKolchanov** [comment]
On my Mac OS Big Sur installation I have found scope.sh and scope.sh.orig in the dir /Library/Python/3.8/site-packages/ranger/data. I replaced scope.sh with scope.sh.orig and now ranger works OK for me. Ranger v1.9.3, Python 3.8.2. pip3 installation.

**Comment by duqcyxwd** [comment]
Yeah, same issue here. It freezes on a vim file with size 1.15k... I guess it is nothing todo with file type.

**Comment by toonn** [comment]
@SergeyKolchanov, could you provide a diff between those files? That might be enlightening.

**Comment by SergeyKolchanov** [comment]
> @SergeyKolchanov, could you provide a diff between those files? That might be enlightening.

Unfortunately, I've lost my original 'bad' scope.sh. My current 'good' one is here

[scope.sh.txt](https://github.com/ranger/ranger/files/6566880/scope.sh.txt)

**Comment by marcelolaia** [comment]
Here is my bad 
[scope.sh.bad.txt](https://github.com/ranger/ranger/files/6570196/scope.sh.bad.txt)

There are differences between then.

I hope this help!

**Comment by michalbrz** [comment]
I can confirm that replacing scope.sh with the one that @SergeyKolchanov shared fixes the issue (thanks!) - for me it was easily reproducible with text files. It removes syntax highlighting though, so it's not perfect 😞

**Comment by akharrou** [comment]
for kitty, i've found that doing:

```zsh
$ kitty @ launch --type=<overlay|window|tab|...> --cwd=current zsh --login
```

fixes everything:
- previewing does freeze ranger
- image previewing works great
- opening images works great (if you've setup `rifle.conf` correctly)
- when you spawn a shell, if you use `--cwd=current`, you'll be in the last visited directory in ranger

```zsh
#: $KITTY_CONFIG_DIRECTORY/kitty.conf

    #: open kitty open in ranger program
    startup_session session/ranger.conf

    #: launch new [zsh] shell overlay; overlays the ranger explorer with a login shell, at the last visited ranger directory
    map super+. launch --type=overlay --cwd=current zsh --login
    # map super+. launch --type=overlay --cwd=current sh -c '/usr/local/bin/gls -Flah --color=auto --hyperlink=auto --time-style=long-iso && exec zsh --login'

    #: launch new file explorer tab; open terminal tab, with ranger explorer opened
    map super+b launch --type=tab --location=neighbor --cwd=current zsh -c ranger
    map shift+super+b launch --type=tab --location=before --cwd=current zsh -c ranger

    #: launch new file explorer [overlay|regular] window, either overlaid/on-top-of the current running program, or to the side as a window pane (you could also use `--type=<hsplit|vsplit>` if you wanted)
    map super+. launch --type=overlay --cwd=current zsh --login
    map super+r launch --type=window --cwd=current ranger
    map shift+super+r launch --type=window --location=before --cwd=current ranger
```

```zsh
#: $KITTY_CONFIG_DIRECTORY/sessions/ranger.conf
new_tab Home
launch ranger
focus
new_tab
launch zsh --login
```

**Comment by vitkabele** [comment]
Hello,
I also got quite annoyed by this freezing issue. My debugging revealed that the issue is on the [ranger/core/loader.py:226](https://github.com/ranger/ranger/blob/master/ranger/core/loader.py#L226) line, which further suggests that the #2502 really fixes the problem. Indeed, after applying patch from there, the behaviour becomes ok.

**Comment by andrewferrier** [comment]
@vitkabele indeed, I have been running with that patch for a little while now and it seems to have cleared up this issue completely. Would be good to see that fix in mainline ranger...

**Comment by detj** [comment]
Freezes regularly when quickly sifting through files using `j/k`. I don't use image/pdf/binary previews. For image, preview just shows the output of `file` utility. I use `bat` for mostly source code text files previewing.
When ranger freezes, `ctrl+c` doesn't work for me. So, I have to find & kill the python process or close my terminal tab/window to recover.
Occasionally, I do get the never ending loading spinner and the corresponding `Not quitting: Task in progress...` message in the status bar. Though, I don't have clear reproducibility on this.

I am running macOS 13.2.1, ranger 1.9.3, bat 0.22.1

**Comment by unphased** [comment]
> I fixed this by setting max lines read in bat using `bat -r 40` during preview text files.

That looks like a kludge to work around the likely root cause of improper process output read as pointed out in previous comments.

**Comment by Bellavene** [comment]
> I wonder if adding preview delay (as `set w3m_delay`) for kitty and iTerm2 will fix the issue

This is the best solution for half of those problems. Ranger's preview mechanism hangs on files with a lot of lines and that is an understandable thing. Do we need those lines viewed when we are seeking for a file pressing down? No. Ranger just tries to make a useless work. So a delay, to cut off repeated keys is a must have to keep our CPU resources intact for useless calculations.
